File: src/main/java/com/morph/engine/script/ScriptSystem.java
package com.morph.engine.script;

import com.morph.engine.core.Game;
import com.morph.engine.core.GameSystem;
import com.morph.engine.entities.Entity;

/**
 * Created on 7/5/2017.
 */
public class ScriptSystem extends GameSystem {
    public ScriptSystem(Game game) {
        super(game);
    }

    @Override
    protected boolean acceptEntity(Entity e) {
        return e.hasComponent(ScriptContainer.class);
    }

    @Override
    public void initSystem() {

    }

    @Override
    protected void preUpdate(Entity e) {
        super.preUpdate(e);

        e.getComponent(ScriptContainer.class).getBehaviors().forEach(EntityBehavior::preUpdate);
    }

    @Override
    protected void update(Entity e) {
        super.update(e);

        e.getComponent(ScriptContainer.class).getBehaviors().forEach(EntityBehavior::update);
    }

    @Override
    protected void fixedUpdate(Entity e, float dt) {
        super.fixedUpdate(e, dt);

        e.getComponent(ScriptContainer.class).getBehaviors().forEach(s -> s.fixedUpdate(dt));
    }

    @Override
    protected void postUpdate(Entity e) {
        super.postUpdate(e);

        e.getComponent(ScriptContainer.class).getBehaviors().forEach(EntityBehavior::postUpdate);
    }
}

File: src/main/kotlin/com/morph/engine/script/ScriptSystem.kt
package com.morph.engine.script

import com.morph.engine.core.Game
import com.morph.engine.core.GameSystem
import com.morph.engine.entities.Entity

/**
 * Created on 7/5/2017.
 */
class ScriptSystem(game: Game) : GameSystem(game) {

    override fun acceptEntity(e: Entity): Boolean {
        return e.hasComponent(ScriptContainer::class.java)
    }

    override fun initSystem() {

    }

    override fun preUpdate(e: Entity) {
        super.preUpdate(e)

        e.getComponent(ScriptContainer::class.java)?.getBehaviors()?.forEach { it.preUpdate() }
    }

    override fun update(e: Entity) {
        super.update(e)

        e.getComponent(ScriptContainer::class.java)?.getBehaviors()?.forEach { it.update() }
    }

    override fun fixedUpdate(e: Entity, dt: Float) {
        super.fixedUpdate(e, dt)

        e.getComponent(ScriptContainer::class.java)?.getBehaviors()?.forEach { it.fixedUpdate(dt) }
    }

    override fun postUpdate(e: Entity) {
        super.postUpdate(e)

        e.getComponent(ScriptContainer::class.java)?.getBehaviors()?.forEach { it.postUpdate() }
    }
}
